cv32e40s/tests/programs/custom/debug_priv_test/debug_priv_test.c

```diff
@@ -0,0 +1,260 @@
+  if (actual != expect) {
+    printf(msg);
+    printf("expected = 0x%lx (%ld), got = 0x%lx (%ld)\n", expect, (int32_t)expect, actual, (int32_t)actual);
+    exit(EXIT_FAILURE);
+  }
+}
+
+/* 
+Retuns specific bit-field from [bit_indx_low : bit_indx_high] in register x
+*/
+unsigned int get_field(unsigned int x, int bit_indx_low, int bit_indx_high){
+    int field = ( 1 << ( (bit_indx_high - bit_indx_low) + 1) )  - 1;
+    x = (x & (field << bit_indx_low) ) >> bit_indx_low;
+    return x;
+}
+
+
+/*
+Checks the mepc for compressed instructions and increments appropriately 
+*/
+void increment_mepc(void){
+  volatile unsigned int insn, mepc;
+
+    __asm__ volatile("csrrs %0, mepc, x0" : "=r"(mepc)); // read the mepc
+    __asm__ volatile("lw %0, 0(%1)" : "=r"(insn) : "r"(mepc)); // read the contents of the mepc pc.
+    if ((insn & 0x3) == 0x3) { // check for compressed instruction before increment.
+      mepc += 4;
+    } else {
+      mepc += 2;
+    }
+    __asm__ volatile("csrrw x0, mepc, %0" :: "r"(mepc)); // write to the mepc 
+}
+
+
+// Rewritten interrupt handler
+__attribute__ ((interrupt ("machine")))
+void u_sw_irq_handler(void) {
+
+    __asm__ volatile("csrrs %0, mcause, x0" : "=r"(wmcause));
+    __asm__ volatile("csrrs %0, mstatus, x0" : "=r"(wmstatus));
+    __asm__ volatile("csrrw x0, mstatus, %0" :: "r"(MSTATUS_STD_VAL)); // set machine mode 
+    increment_mepc();
+}
+
+// Declaration of the debug struct 
+typedef union {
+  struct {
+    unsigned int start_delay      : 15; // 14: 0
+    unsigned int rand_start_delay : 1;  //    15
+    unsigned int pulse_width      : 13; // 28:16
+    unsigned int rand_pulse_width : 1;  //    29
+    unsigned int pulse_mode       : 1;  //    30    0 = level, 1 = pulse
+    unsigned int value            : 1;  //    31
+  } fields;
+  unsigned int bits;
+}  debug_req_control_t;
+
+/*
+When called it starts debug_mode
+*/
+void run_debug_mode(void) {
+  // debug control struct 
+  debug_req_control_t debug_req_control;
+  debug_req_control = (debug_req_control_t) {
+    .fields.value            = 1,
+    .fields.pulse_mode       = 1, //PULSE Mode
+    .fields.rand_pulse_width = 0,
+    .fields.pulse_width      = 5,// FIXME: BUG: one clock pulse cause core to lock up
+    .fields.rand_start_delay = 0,
+    .fields.start_delay      = 200
+  };
+
+  // this will initiate debug mode 
+  DEBUG_REQ_CONTROL_REG = debug_req_control.bits; 
+  // waits for debug to finish
+  while(glb_debug_status != 1){
+      ;
+  }
+  // resets the debug status bit
+  glb_debug_status = 0;
+}
+
+int main(void){// TODO: test will failed until issue #278 in core-v-verif/cv32e40s is fixed
+  setup_pmp();
+
+
+
+
+  // Test start:
+  /*
+  Have dcsr.ebreaku=0, be in U-mode, execute ebreak, ensure "normal" ebreak behavior and no debug entry.
+  */
+  set_u_mode();
+  asm volatile("ebreak");
+  assert_or_die(wmcause, 0x3, "Error: Illegal 'ebreak' did not trigger breakpoint exception!");
```

Comment:
Would it make sense to also check `glb_debug_status`?

Reply:
It could. Although, `glb_debug_status` should be 0 at all times, except the time in-between debug exit and before `run_debug_mode()` exits. The only other case I see is if for some other reason the core is in debug mode before main() executes, but I don't know if this is applicable? And if it is then I don't know if my version of the `debugger_start:` symbol has loaded as the core would already be using some version of this symbol when it entered debug mode. 

In any case I don't see any issues with adding an `assert_or_die` case checking that the `glb_debug_status` but is low.

Reply:
> Although, glb_debug_status should be 0 at all times, except the time in-between debug exit and before run_debug_mode() exits.

If the ebreak mistakenly caused debug mode to be entered, would `glb_debug_status` be set then?

Reply:
In a way, but only until the debug assembly code exits, and right after the `run_debug_mode()` function will set `glb_debug_status` to `0` again so that it has something to compare to next time its called. The way I've rewritten it `glb_debug_status` isn't meant to be a check for whether debug mode is enabled, its there to tell the `run_debug_mode()` that the debug assembly code has finished executing and that the .c code can now move on with testing. This exists because it seemed to be the case for the other custom test in the cv32e40s called "debug_test" where the executive .c code seems to wait for the assembly program to finish before moving on.